src/hardwareModel.py

```diff
@@ -330,3 +341,46 @@ def get_mem_compute_bw(self):
         edges = list(filter(lambda x: "Reg" in x[0], self.netlist.in_edges("Buf0")))
         n = len(edges)
         return n
+    
+    def gen_cacti_results(self):
+
+        # 1. add bit width, mem size, cache size DONE
+        # 2. add IO DONE
+        # 3. check simulate.py for energy instead of power ASK about this
+
+        # 1 details
+        # sizes mem set on hwmodel memsize var, add variable for cache size in hardWareModel that we set later - init small
+        # add instance for all -> cache size, bit width
+
+        # 2 details
+        # add I/O 
+        # latency and power to each main mem read or write
+        # add new dict entry, off-chip IO [latency and power], add that whenever mem interaction DONE
+
+        # 3 details
+        # careful cacti gen energy 
+        # need to change power to energy -> just for buf and main mem DONE 
+        # add dynamic energy, change simulate.py  
+        # store read and write energy separate, buf read & write -> dynamic energy, key for read and write DONE
+
+        buf_vals = gen_vals("base_cache", 131072, 64,
+                                      "cache", 512)
+        mem_vals = gen_vals("mem_cache", 131072, 64,
+                                      "main memory", 512)
```

Comment:
Two small changes. 
the bus width here should be set from the `hw.bus_width` variable that you created.
And, a parameter to this function should be which technology library to use.
hardwareModel has a `transistor_size` variable, and cacti has a list of which technology libraries it uses. need to add a function that maps `transistor_size` to the nearest allowed technology library.